File: servers/server_clear_thought/tools/analogical_mapper.py
from __future__ import annotations

from typing import Any, Dict, List

from pydantic import BaseModel

from ..core.base_tool import BaseTool


class AnalogicalMapper(BaseTool):
    slug = "analogical_mapper"
    endpoint_path = "/analogical-mapper"

    class InputSchema(BaseModel):
        problem: str
        seed_domains: List[str] | None = None
        k: int | None = 3

    class OutputSchema(BaseModel):
        analogies: List[Dict[str, Any]]
        suggested_prompts: List[str]

    def execute(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        k = payload.get("k") or 3
        analogies = [
            {"domain": d, "analogy": f"{payload['problem']} ~ {d}"}
            for d in (payload.get("seed_domains") or ["math", "biology", "art"])[:k]
        ]
        prompts = [f"How would {a['domain']} approach it?" for a in analogies]
        return {
            "analogies": analogies,
            "suggested_prompts": prompts,
        }

File: servers/server-clear-thought/src/tools/analogical-mapper.ts
import { z } from 'zod';
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import type { SessionState } from '../state/SessionState.js';

export function registerAnalogicalMapper(server: McpServer, _sessionState: SessionState) {
  server.tool(
    'analogical_mapper',
    'Generate analogies from seed domains and suggest prompts',
    {
      problem: z.string(),
      seed_domains: z.array(z.string()).optional(),
      k: z.number().int().optional()
    },
    async (args) => {
      const k = args.k ?? 3;
      const domains = (args.seed_domains ?? ['math', 'biology', 'art']).slice(0, k);
      const analogies = domains.map((d) => ({ domain: d, analogy: `${args.problem} ~ ${d}` }));
      const suggested_prompts = analogies.map((a) => `How would ${a.domain} approach it?`);
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify({ analogies, suggested_prompts }, null, 2)
          }
        ]
      };
    }
  );
}
